File: engine/charset.go
package engine

import "unicode"

// Diacritic types
const (
	DiacriticNone       = 0
	DiacriticBreve      = 1 // ă
	DiacriticCircumflex = 2 // â ê ô
	DiacriticHorn       = 3 // ơ ư
	DiacriticStroke     = 4 // đ
)

// Tone types: 0 none, 1 sắc, 2 huyền, 3 hỏi, 4 ngã, 5 nặng
const (
	ToneNone  = 0
	ToneSac   = 1 // s / 1
	ToneHuyen = 2 // f / 2
	ToneHoi   = 3 // r / 3
	ToneNga   = 4 // x / 4
	ToneNang  = 5 // j / 5
)

type CharInfo struct {
	Bare      rune // 'a','e','i','o','u','y','d' lower
	Diacritic int
	Tone      int
	IsUpper   bool
}

var charInfoMap map[rune]CharInfo
var reverseMap map[CharInfo]rune

func init() {
	charInfoMap = make(map[rune]CharInfo)
	reverseMap = make(map[CharInfo]rune)

	// Helper to register
	register := func(r rune, bare rune, dia, tone int, upper bool) {
		info := CharInfo{Bare: bare, Diacritic: dia, Tone: tone, IsUpper: upper}
		charInfoMap[r] = info
		reverseMap[info] = r
	}

	// Define groups: each string first char is base without tone, following 5 are tones 1-5
	// Lowercase groups
	lowerGroups := []struct {
		bare rune
		dia  int
		chars string // len 6: base + 5 toned
	}{
		{'a', DiacriticNone, "aáàảãạ"},
		{'a', DiacriticBreve, "ăắằẳẵặ"},
		{'a', DiacriticCircumflex, "âấầẩẫậ"},
		{'e', DiacriticNone, "eéèẻẽẹ"},
		{'e', DiacriticCircumflex, "êếềểễệ"},
		{'i', DiacriticNone, "iíìỉĩị"},
		{'o', DiacriticNone, "oóòỏõọ"},
		{'o', DiacriticCircumflex, "ôốồổỗộ"},
		{'o', DiacriticHorn, "ơớờởỡợ"},
		{'u', DiacriticNone, "uúùủũụ"},
		{'u', DiacriticHorn, "ưứừửữự"},
		{'y', DiacriticNone, "yýỳỷỹỵ"},
	}
	for _, g := range lowerGroups {
		rs := []rune(g.chars)
		for idx, r := range rs {
			register(r, g.bare, g.dia, idx, false) // idx is tone
		}
	}
	// Uppercase groups
	upperGroups := []struct {
		bare rune
		dia  int
		chars string
	}{
		{'a', DiacriticNone, "AÁÀẢÃẠ"},
		{'a', DiacriticBreve, "ĂẮẰẲẴẶ"},
		{'a', DiacriticCircumflex, "ÂẤẦẨẪẬ"},
		{'e', DiacriticNone, "EÉÈẺẼẸ"},
		{'e', DiacriticCircumflex, "ÊẾỀỂỄỆ"},
		{'i', DiacriticNone, "IÍÌỈĨỊ"},
		{'o', DiacriticNone, "OÓÒỎÕỌ"},
		{'o', DiacriticCircumflex, "ÔỐỒỔỖỘ"},
		{'o', DiacriticHorn, "ƠỚỜỞỠỢ"},
		{'u', DiacriticNone, "UÚÙỦŨỤ"},
		{'u', DiacriticHorn, "ƯỨỪỬỮỰ"},
		{'y', DiacriticNone, "YÝỲỶỸỴ"},
	}
	for _, g := range upperGroups {
		rs := []rune(g.chars)
		for idx, r := range rs {
			register(r, g.bare, g.dia, idx, true)
		}
	}
	// Stroke đ/Đ (no tone)
	register('d', 'd', DiacriticNone, ToneNone, false)
	register('đ', 'd', DiacriticStroke, ToneNone, false)
	register('D', 'd', DiacriticNone, ToneNone, true)
	register('Đ', 'd', DiacriticStroke, ToneNone, true)

	// Also register plain consonants as themselves? For isVowel check we use map
}

func lookup(bare rune, dia, tone int, isUpper bool) (rune, bool) {
	info := CharInfo{Bare: bare, Diacritic: dia, Tone: tone, IsUpper: isUpper}
	r, ok := reverseMap[info]
	return r, ok
}

// isVowel checks if rune is Vietnamese vowel (including toned variants)
func isVowel(r rune) bool {
	info, ok := charInfoMap[r]
	if !ok {
		return false
	}
	// đ is not vowel even though in map
	if info.Bare == 'd' {
		return false
	}
	return true
}

func isConsonant(r rune) bool {
	if isVowel(r) {
		return false
	}
	// letter and not vowel => consonant (including đ? đ considered consonant)
	if unicode.IsLetter(r) {
		return true
	}
	return false
}

func isLetter(r rune) bool { return unicode.IsLetter(r) }

func toLower(r rune) rune { return unicode.ToLower(r) }

func applyToneAt(word []rune, pos int, tone int) []rune {
	if pos < 0 || pos >= len(word) {
		return word
	}
	r := word[pos]
	info, ok := charInfoMap[r]
	if !ok {
		return word
	}
	if info.Bare == 'd' {
		return word
	}
	// Preserve diacritic, change tone
	newRune, ok := lookup(info.Bare, info.Diacritic, tone, info.IsUpper)
	if !ok {
		return word
	}
	word[pos] = newRune
	return word
}

func removeAllTones(word []rune) []rune {
	for i, r := range word {
		info, ok := charInfoMap[r]
		if !ok {
			continue
		}
		if info.Tone != ToneNone {
			if nr, ok := lookup(info.Bare, info.Diacritic, ToneNone, info.IsUpper); ok {
				word[i] = nr
			}
		}
	}
	return word
}

func toggleBreve(r rune) (rune, bool) {
	info, ok := charInfoMap[r]
	if !ok {
		return r, false
	}
	if info.Bare != 'a' {
		return r, false
	}
	var newDia int
	if info.Diacritic == DiacriticBreve {
		newDia = DiacriticNone
	} else if info.Diacritic == DiacriticNone || info.Diacritic == DiacriticCircumflex {
		newDia = DiacriticBreve
	} else {
		return r, false
	}
	nr, ok := lookup(info.Bare, newDia, info.Tone, info.IsUpper)
	if !ok {
		return r, false
	}
	return nr, true
}

func toggleCircumflex(r rune) (rune, bool) {
	info, ok := charInfoMap[r]
	if !ok {
		return r, false
	}
	var newDia int
	switch info.Bare {
	case 'a':
		if info.Diacritic == DiacriticCircumflex {
			newDia = DiacriticNone
		} else if info.Diacritic == DiacriticNone || info.Diacritic == DiacriticBreve {
			newDia = DiacriticCircumflex
		} else {
			return r, false
		}
	case 'e':
		if info.Diacritic == DiacriticCircumflex {
			newDia = DiacriticNone
		} else if info.Diacritic == DiacriticNone {
			newDia = DiacriticCircumflex
		} else {
			return r, false
		}
	case 'o':
		if info.Diacritic == DiacriticCircumflex {
			newDia = DiacriticNone
		} else if info.Diacritic == DiacriticNone || info.Diacritic == DiacriticHorn {
			newDia = DiacriticCircumflex
		} else {
			return r, false
		}
	default:
		return r, false
	}
	nr, ok := lookup(info.Bare, newDia, info.Tone, info.IsUpper)
	if !ok {
		return r, false
	}
	return nr, true
}

func toggleHorn(r rune) (rune, bool) {
	info, ok := charInfoMap[r]
	if !ok {
		return r, false
	}
	var newDia int
	switch info.Bare {
	case 'o':
		if info.Diacritic == DiacriticHorn {
			newDia = DiacriticNone
		} else if info.Diacritic == DiacriticNone || info.Diacritic == DiacriticCircumflex {
			newDia = DiacriticHorn
		} else {
			return r, false
		}
	case 'u':
		if info.Diacritic == DiacriticHorn {
			newDia = DiacriticNone
		} else if info.Diacritic == DiacriticNone {
			newDia = DiacriticHorn
		} else {
			return r, false
		}
	default:
		return r, false
	}
	nr, ok := lookup(info.Bare, newDia, info.Tone, info.IsUpper)
	if !ok {
		return r, false
	}
	return nr, true
}

func toggleStroke(r rune) (rune, bool) {
	info, ok := charInfoMap[r]
	if !ok {
		return r, false
	}
	if info.Bare != 'd' {
		return r, false
	}
	var newDia int
	if info.Diacritic == DiacriticStroke {
		newDia = DiacriticNone
	} else {
		newDia = DiacriticStroke
	}
	nr, ok := lookup(info.Bare, newDia, ToneNone, info.IsUpper)
	if !ok {
		return r, false
	}
	return nr, true
}

func getTone(r rune) int {
	info, ok := charInfoMap[r]
	if !ok {
		return ToneNone
	}
	return info.Tone
}

func getDiacritic(r rune) int {
	info, ok := charInfoMap[r]
	if !ok {
		return DiacriticNone
	}
	return info.Diacritic
}

func isUpper(r rune) bool { return unicode.IsUpper(r) }

// helper to get bare lower
func bareLower(r rune) rune {
	info, ok := charInfoMap[r]
	if !ok {
		return toLower(r)
	}
	return info.Bare
}

File: engine/charset.c
#include "internal.h"

/* Global maps */
static GHashTable *char_info_map = NULL; // gunichar (as gpointer) -> CharInfo*
static GHashTable *reverse_map = NULL;   // key string "bare:dia:tone:upper" -> gunichar

gchar* reverse_key(gunichar bare, int dia, int tone, gboolean upper){
    return g_strdup_printf("%u:%d:%d:%d", (guint)bare, dia, tone, upper?1:0);
}

void register_char(gunichar ch, gunichar bare, int dia, int tone, gboolean upper){
    CharInfo *info = g_new(CharInfo,1);
    info->bare=bare; info->diacritic=dia; info->tone=tone; info->is_upper=upper;
    g_hash_table_insert(char_info_map, GUINT_TO_POINTER((guint)ch), info);
    gchar *k = reverse_key(bare,dia,tone,upper);
    g_hash_table_insert(reverse_map, k, GUINT_TO_POINTER((guint)ch));
}

void init_charset(void){
    if (char_info_map) return;
    char_info_map = g_hash_table_new(g_direct_hash, g_direct_equal);
    reverse_map = g_hash_table_new_full(g_str_hash, g_str_equal, g_free, NULL);
    // lower groups - strings UTF-8
    struct { gunichar bare; int dia; const char *utf8; } lower[] = {
        {'a', DIAC_NONE, "aáàảãạ"},
        {'a', DIAC_BREVE, "ăắằẳẵặ"},
        {'a', DIAC_CIRCUMFLEX, "âấầẩẫậ"},
        {'e', DIAC_NONE, "eéèẻẽẹ"},
        {'e', DIAC_CIRCUMFLEX, "êếềểễệ"},
        {'i', DIAC_NONE, "iíìỉĩị"},
        {'o', DIAC_NONE, "oóòỏõọ"},
        {'o', DIAC_CIRCUMFLEX, "ôốồổỗộ"},
        {'o', DIAC_HORN, "ơớờởỡợ"},
        {'u', DIAC_NONE, "uúùủũụ"},
        {'u', DIAC_HORN, "ưứừửữự"},
        {'y', DIAC_NONE, "yýỳỷỹỵ"},
        {0,0,NULL}
    };
    for(int i=0; lower[i].utf8; i++){
        glong len;
        gunichar *ucs = g_utf8_to_ucs4(lower[i].utf8, -1, NULL, &len, NULL);
        for(int t=0; t<len; t++){
            register_char(ucs[t], lower[i].bare, lower[i].dia, t, FALSE);
        }
        g_free(ucs);
    }
    struct { gunichar bare; int dia; const char *utf8; } upper[] = {
        {'a', DIAC_NONE, "AÁÀẢÃẠ"},
        {'a', DIAC_BREVE, "ĂẮẰẲẴẶ"},
        {'a', DIAC_CIRCUMFLEX, "ÂẤẦẨẪẬ"},
        {'e', DIAC_NONE, "EÉÈẺẼẸ"},
        {'e', DIAC_CIRCUMFLEX, "ÊẾỀỂỄỆ"},
        {'i', DIAC_NONE, "IÍÌỈĨỊ"},
        {'o', DIAC_NONE, "OÓÒỎÕỌ"},
        {'o', DIAC_CIRCUMFLEX, "ÔỐỒỔỖỘ"},
        {'o', DIAC_HORN, "ƠỚỜỞỠỢ"},
        {'u', DIAC_NONE, "UÚÙỦŨỤ"},
        {'u', DIAC_HORN, "ƯỨỪỬỮỰ"},
        {'y', DIAC_NONE, "YÝỲỶỸỴ"},
        {0,0,NULL}
    };
    for(int i=0; upper[i].utf8; i++){
        glong len;
        gunichar *ucs = g_utf8_to_ucs4(upper[i].utf8, -1, NULL, &len, NULL);
        for(int t=0; t<len; t++){
            register_char(ucs[t], upper[i].bare, upper[i].dia, t, TRUE);
        }
        g_free(ucs);
    }
    register_char('d', 'd', DIAC_NONE, TONE_NONE, FALSE);
    register_char(0x0111, 'd', DIAC_STROKE, TONE_NONE, FALSE); // đ
    register_char('D', 'd', DIAC_NONE, TONE_NONE, TRUE);
    register_char(0x0110, 'd', DIAC_STROKE, TONE_NONE, TRUE); // Đ
}

CharInfo* get_info(gunichar c){
    return (CharInfo*)g_hash_table_lookup(char_info_map, GUINT_TO_POINTER((guint)c));
}
gboolean lookup_char(gunichar bare, int dia, int tone, gboolean upper, gunichar *out){
    gchar *k = reverse_key(bare,dia,tone,upper);
    gpointer v = g_hash_table_lookup(reverse_map, k);
    g_free(k);
    if (!v) return FALSE;
    *out = (gunichar)GPOINTER_TO_UINT(v);
    return TRUE;
}
gboolean is_vowel(gunichar c){
    CharInfo *info = get_info(c);
    if (!info) return FALSE;
    if (info->bare == 'd') return FALSE;
    return TRUE;
}
gboolean is_consonant(gunichar c){
    if (is_vowel(c)) return FALSE;
    return g_unichar_isalpha(c);
}
gunichar to_lower_g(gunichar c){ return g_unichar_tolower(c); }
gint get_tone(gunichar c){ CharInfo *i=get_info(c); return i?i->tone:TONE_NONE; }
gint get_diac(gunichar c){ CharInfo *i=get_info(c); return i?i->diacritic:DIAC_NONE; }
gunichar bare_lower(gunichar c){ CharInfo *i=get_info(c); if(i) return i->bare; return to_lower_g(c); }

gboolean toggle_breve(gunichar in, gunichar *out){
    CharInfo *info=get_info(in); if(!info||info->bare!='a') return FALSE;
    int ndia=-1;
    if(info->diacritic==DIAC_BREVE) ndia=DIAC_NONE;
    else if(info->diacritic==DIAC_NONE || info->diacritic==DIAC_CIRCUMFLEX) ndia=DIAC_BREVE;
    else return FALSE;
    return lookup_char(info->bare, ndia, info->tone, info->is_upper, out);
}
gboolean toggle_circumflex(gunichar in, gunichar *out){
    CharInfo *info=get_info(in); if(!info) return FALSE;
    int ndia=-1;
    if(info->bare=='a' || info->bare=='e' || info->bare=='o'){
        if(info->diacritic==DIAC_CIRCUMFLEX) ndia=DIAC_NONE;
        else if(info->diacritic==DIAC_NONE || info->diacritic==DIAC_BREVE) ndia=DIAC_CIRCUMFLEX;
        else return FALSE;
        return lookup_char(info->bare, ndia, info->tone, info->is_upper, out);
    }
    return FALSE;
}
gboolean toggle_horn(gunichar in, gunichar *out){
    CharInfo *info=get_info(in); if(!info) return FALSE;
    if(info->bare!='o' && info->bare!='u') return FALSE;
    int ndia=-1;
    if(info->diacritic==DIAC_HORN) ndia=DIAC_NONE;
    else if(info->diacritic==DIAC_NONE || info->diacritic==DIAC_CIRCUMFLEX) ndia=DIAC_HORN;
    else return FALSE;
    return lookup_char(info->bare, ndia, info->tone, info->is_upper, out);
}
gboolean toggle_stroke(gunichar in, gunichar *out){
    CharInfo *info=get_info(in); if(!info||info->bare!='d') return FALSE;
    int ndia = (info->diacritic==DIAC_STROKE)?DIAC_NONE:DIAC_STROKE;
    return lookup_char(info->bare, ndia, TONE_NONE, info->is_upper, out);
}

/* Phonology: find tone position */
